catalog_persistence/databases.py

```diff
@@ -0,0 +1,295 @@
+import abc
+from datetime import datetime
+from enum import Enum
+from uuid import uuid4
+
+import couchdb
+
+from catalog_persistence.models import DocumentRecord
+
+
+class DocumentNotFound(Exception):
+    pass
+
+
+class ChangeType(Enum):
+    CREATE = 'C'
+    UPDATE = 'U'
+    DELETE = 'D'
+
+
+class Change:
+
+    def __init__(self, document_record, change_type):
+        self._record_id = uuid4().hex
+        self._document_id = document_record.document_id
+        self._document_type = document_record.document_type.value
+        self._type = change_type.value
+
+    @property
+    def document_id(self):
+        return self._record_id
+
+    @document_id.setter
+    def document_id(self, record_id):
+        self._record_id = record_id
+
+    @property
+    def created_date(self):
+        return self._created_date
+
+    @created_date.setter
+    def created_date(self, created_date):
+        self._created_date = created_date
+
+    def input(self):
+        return {
+            'record_id': self.document_id,
+            'document_id': self._document_id,
+            'document_type': self._document_type,
+            'type': self._type,
+            'created_date': self.created_date,
+        }
+
+    def output(self):
+        return {}
+
+
+class BaseDBManager(metaclass=abc.ABCMeta):
+
+    @abc.abstractmethod
+    def database(self, database_name) -> None:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def drop_database(self) -> None:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def create(self, document) -> str:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def read(self, id) -> dict:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def update(self, document) -> str:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def delete(self, id) -> None:
+        return NotImplemented
+
+    @abc.abstractmethod
+    def find(self) -> list:
+        return NotImplemented
+
+
+class InMemoryDBManager(BaseDBManager):
+
+    def __init__(self, config):
+        self._db_name = None
+        self._database = {}
+
+    @property
+    def database_name(self):
+        return self._db_name
+
+    @property
+    def database(self):
+        table = self._database.get(self._db_name)
+        if not table:
+            self._database[self._db_name] = {}
+        return self._database[self._db_name]
+
+    @database.setter
+    def database(self, database_name):
+        self._db_name = database_name
+
+    def drop_database(self):
+        self._database = {}
+
+    def create(self, document):
+        document.created_date = str(datetime.utcnow().timestamp())
```

Comment:
Tudo bem que trata-se da implementação in-memory apenas para testes, mas aqui há uma dependência que inviabiliza qualquer tipo de injeção para isolar a dependência ao pacote _datetime_.